gusto/recovery.py

```diff
@@ -163,91 +167,91 @@ def __init__(self, v0, v1, v_out, method='density'):
             max_coord = Function(VDG0).interpolate(Constant(np.max(self.coords.dat.data[:, 0])))
 
             right_kernel = """
-            if (fmax(COORDS[0][0], fmax(COORDS[1][0], COORDS[2][0])) == MAX[0][0])
-            RIGHT[0][0] = 1.0;
+            if (fmax(COORDS[0], fmax(COORDS[2], COORDS[2*2])) == MAX[0])
+                RIGHT[0] = 1.0;
             """
             par_loop(right_kernel, dx,
                      args={"COORDS": (self.coords, READ),
                            "MAX": (max_coord, READ),
-                           "RIGHT": (self.right, RW)})
+                           "RIGHT": (self.right, WRITE)})
 
             self.bottom_kernel = """
-            if (RIGHT[0][0] == 1.0)
+            if (RIGHT[0] == 1.0)
             {
-            float x = COORDS[2][0] - COORDS[0][0];
-            float y = COORDS[1][1] - COORDS[0][1];
-            float a = CG1[3][0];
-            float b = CG1[1][0];
-            float c = DG0[0][0];
-            DG1[1][0] = a;
-            DG1[3][0] = b;
-            DG1[2][0] = (1.0 / (pow(x, 2.0) + 4.0 * pow(y, 2.0))) * (-3.0 * a * pow(y, 2.0) - b * pow(x, 2.0) - b * pow(y, 2.0) + 2.0 * c * pow(x, 2.0) + 8.0 * c * pow(y, 2.0));
-            DG1[0][0] = 4.0 * c - b - a - DG1[2][0];
+            float x = COORDS[2*2] - COORDS[0];
+            float y = COORDS[2*1 + 1] - COORDS[1];
```

Comment:
To make sure I've got this right, for vector function `f` in spatial dimension `d` we have `f[a][b] -> f[a*d + b]` ?

Reply:
Yes.